modules/idlgen/core/src/main/scala/ScalaParser.scala

```diff
@@ -14,43 +14,75 @@
  * limitations under the License.
  */
 
-package freestyle.rpc.idlgen
+package freestyle.rpc
+package idlgen
 
-import freestyle.rpc.internal.ServiceAlg
-import freestyle.rpc.internal.util.ScalametaUtil._
 import freestyle.rpc.internal.util.StringUtil._
-import scala.meta._
+import freestyle.rpc.internal.util.{AstOptics, Toolbox}
+import freestyle.rpc.protocol._
 
 object ScalaParser {
 
-  def parse(input: Source, inputName: String): RpcDefinitions = {
-    val definitions = input.collect { case defn: Defn => defn }
+  import Toolbox.u._
+  import AstOptics._
+  import Model._
 
-    // format: OFF
-    def annotationValue(name: String): Option[String] = (for {
-      defn <- definitions
-      annotation <- defn.annotationsNamed(name)
-      firstArg <- annotation.firstArg
-    } yield firstArg).headOption.map(_.toString.unquoted)
+  def parse(
+      input: Tree,
+      inputName: String
+  ): RpcDefinitions = {
 
-    val outputName = annotationValue("outputName").getOrElse(inputName)
+    val definitions = input.collect { case defs: ModuleDef => defs }
+
+    def annotationValue(name: String): Option[String] =
+      (for {
+        defn       <- definitions
+        annotation <- annotationsNamed(name).getAll(defn)
+        firstArg   <- annotation.firstArg
+      } yield firstArg).headOption.map(_.toString.unquoted)
+
+    val outputName    = annotationValue("outputName").getOrElse(inputName)
     val outputPackage = annotationValue("outputPackage")
 
     val options: Seq[RpcOption] = for {
-      defn  <- definitions
-      option <- defn.annotationsNamed("option")
+      defn             <- definitions
+      option           <- annotationsNamed("option").getAll(defn)
       Seq(name, value) <- option.withArgsNamed("name", "value")
     } yield RpcOption(name.toString.unquoted, value.toString) // keep value quoting as-is
 
-    val messages: Seq[RpcMessage] = definitions.filter(_.hasAnnotation("message")).map { defn =>
-      RpcMessage(defn.name, defn.params)
+    val messages: Seq[RpcMessage] = definitions.head.collect {
```

Comment:
`.head` would be unsafe :/

Reply:
yep, good catch! I had notes about fixing all those but forgot :)